src/server/master_server/chunk_server_manager.cc

```diff
@@ -1,69 +1,304 @@
 #include "src/server/master_server/chunk_server_manager.h"
 
+namespace protos {
+bool operator==(const protos::ChunkServerLocation& lhs,
+                const protos::ChunkServerLocation& rhs) {
+  return lhs.server_hostname() == rhs.server_hostname() &&
+         lhs.server_port() == rhs.server_port();
+}
+
+bool operator==(const protos::ChunkServer& lhs,
+                const protos::ChunkServer& rhs) {
+  return lhs.location() == rhs.location();
+}
+}  // namespace protos
+
 namespace gfs {
 
 namespace server {
 
-// Iterates through the chunk servers and selects the server with
-// the max available memory.
-// TODO(bmokutub): optimize this, maybe use a sorted map.
-std::shared_ptr<protos::ChunkServer> ChunkServerManager::AllocateChunkServer() {
-  auto servers_iterator = this->chunk_servers_map_.begin();
+ChunkServerLocationThreadSafeFlatSet ChunkServerManager::AllocateChunkServer(
```

Comment:
I think the return here should not be a thread-safe container per se? Each caller of this function should just get a list of copies of the chunk servers, and these copies are not shared among those callers. That can simplify things quite a bit isn't it? 

Furthermore, given how a single lock is acquired, this thread-safe data structure is not necessary.

Reply:
You're correct, the caller should just get a simple set. Returning threadsafeset here because that is what we are using to store it internally and didn't want to convert it (copying) from simpleset to threadsafe when storing, just for the sake of returning a simple structure. Because it could potentially be a large set. For internal impl, we need threadsafe set because a thread can be adding a location for a chunk_handle and another may be remove a location from it because the chunkserver is down. The caller shouldn't care if it's threadsafe or not and just use the set returned. Don't mind changing it to do the copy. But what do you think?

Reply:
Will change to return a simple set, but will store internally with a threadsafe set because we need that as explained above.

Reply:
From a practical point of view this set won't be too large I think - a chunk is likely to be replicated into 3 or 5 (I guess in the very rare case it can be larger to improve read throughput?). 

But more importantly, I think it is the caller's responsibility to deal with the condition that a chunk is moved away. Because everything happens concurrently, things can come and go (and dealing with deletion is a pain overall). I think it adds complexity if the caller has to be aware of the concurrency of a returned structure when making a query. For instance, based on what I read, iterators in this kind of thread-safe data structure is not safe because of concurrency, and the fact that it stores references adds to it. 

Changing this to a copied list means the caller get a snapshot of what happened in that moment, and will find out about it later using that list. If there is a chunk that is not there, it is the caller's call to deal with it. 

I hope this sounds reasonable? If so let's convert this to just a list of copies, I believe it will simplify things on both end.

Reply:
True, what I said previously was that the caller doesn't need to care about the returned type as threadsafe since they are getting a copy anyway. Was returning that to avoid copying again. Practically the set won't be 3 or 5, since clients set their replication target. But for our prototype sake can assume that. Yeah will copy it to a simple set and return that.

Reply:
On second thought, leaving this like this. I don't see much need in copying back and forth just because I want to simplify the return type. Forexample: when GetChunkLocation is called, We store it internally as a threadsafeset, currently just returning a copy of that. But I don't see much need to then copying content of the threadsafe set into a simpleset then return just because of simplified return type. The caller gets copy of value and can do anything with it, shouldn't be bothered because they got a threadsafe set. If the caller doesn't care, then they use it the same way they would have used a normal set. Happy to chat about this on slack if needed.

Reply:
No callers share any of this ThreadSafeSet, so why do we create a local copy of this an return them? Creating these ThreadSafeSet is not trivial I think. Furthermore, why do we need a set not just a vector?

Reply:
Please kindly take a look at the code for AllocateChunk and GetChunkLocation in chunkservermgr code may be it will help. Also a set is used because we want a unique collection of locations. Don't want to have a location multiple times in the chunk location collection.

Reply:
I agree with your underlying principle here. But not really needed here since as I mentioned there is no much value in copying the values to simplified type just for the sake of returning a simplified type. As I mentioned, caller shouldn't care whether the set is threadsafe or not. Just use the set the way they want. Please if you think I'm being unreasonable here, feel free to paste a code snippet of how you would change this and maybe that will help us have a better discussion and see the perspectives here.

Reply:
Using threadsafe set internally to support concurrent adding/removing to the chunk locations. A thread can be adding loc x to chunk a while another thread is adding loc y to chunk a.

Reply:
Hi Micheal, this type of thread-safe set internally uses subsets and associated locks so I think instantiating them comes with a different cost than just normal set. Furthermore, because the selection algorithm is predicated on a single lock, I don't see how concurrent add / remove happens to this return type? Can we at least just use a normal set for this? I don't think this is just for the sake of returning a simplified type.

Reply:
I have chatted with Michael on the phone and we converge on this issue. It is ok for us to move forward with this ThreadSafeSet as is as it doesn't really affect me as the caller as the returned one is not shared. There are some overhead to convert it to normal set and Michale will leave a TODO in his revision. Let's not be blocked by this.